README
/* vim:set ts=2 nowrap: ****************************************************

 librtdebug - A C++ based thread-safe Runtime Debugging Library
 Copyright (C) 2003-2005 by Jens Langner <Jens.Langner@light-speed.de>

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 $Id$

***************************************************************************/

librtdebug is a small but powerfull debugging purpose library for C++
developers. Its purpose is to provide the developer with some set of
C-language macro definitions which can be easily inserted in all kind of
applications. These macros then call internal class methods of the debugging
library which output the desired data during the execution of the program.

In addition to that it takes care of the correct indention of the output,
providing the developer a nice and easy way to track possible problems due
to execution output at the standard output devices.

Aside from that, it also automatically keeps track if it is called out of
different threads, using different indention levels and other means of
variations to make it easy to recognize if output is comming from the
context of another thread.

Requirements:
------------
As the library is fully written in C++ you at least require a modern, uptodate
C++ compiler, preferably GCC in its latest 3.4+ version.
In addition the build system is done with the qmake utility of the Qt framework
kit, which raises the requirement for a recent Qt3 version to make it compile
out of the box. However, the code itself doesn't require any Qt classes or
link any of them to the library itself. This allows to use librtdebug also
with non Qt applications.

Features:
--------

- small and easy macro based interface
- fully object oriented design and implementation using C++
- thread-safe implementation allowing to track the output of a certain thread
- highly portable implementation with direct implementations existing for
	SunOS, MacOSX and Linux. (more to come)
- Doxygen based class documentation and a deeper implementation 
	documentation based on the MSc thesis of my computer science study.
- Released under LGPL (Lesser General Public License) for a maximum
	available flexibility and developer support aswell as the possibility to
	use the library in commercial applications.

Future plans:
------------
Have a look at the TODO file.

Please feel free to contact me if you have any questions about the
implementation of libmedio and it's underlying algorithms. But please make
sure that you have checked the whole documentation first.

Jens Langner <Jens.Langner@light-speed.de>
January, 2005


ChangeLog
------------------------------------------
RTDebugLib - Runtime Debugging C++ Library
------------------------------------------
// vim:set ts=2 nowrap:

$Id$
$URL$

2015-04-16  Jens Maus <mail@jens-maus.de>

   * misc: switch build environment from qmake to cmake and implemented mxe-based
     cross compiling support for win32.

2014-05-21  Thore Boeckelmann <tboeckel@gmx.de>

   * CRTDebug.cpp: changed the type of the value parameter from to "long" to
     "long long" to allow output of 64bit variables with the SHOWVALUE() macro.

2013-11-08  Thore Boeckelmann <tboeckel@gmx.de>

   * configure: allow gcc versions like 4.6, too, instead of 4.6.x only.
   * CRTDebug.cpp: correctly switch back to decimal output of numbers after
     printing some hexdecimal numbers.
   * CRTDebug.cpp: include the process ID in the output to assist parallel
     usage of librtdebug in multiple applications all of which print to the
     same shell.

2009-07-28  Jens Langner <Jens.Langner@light-speed.de>

   * CRTDebug.c: fixed the broken build when building rtdebug with GCC 4.x.

2008-06-03  Jens Langner <Jens.Langner@light-speed.de>

   * CRTDebug.c: added va_copy() usage to make the new dynamic allocation
     for non-vasprintf() systems more safe.
   * CRTDebug.c: added alternative implementation for the dynamic string
     allocation in case a system does not have any vasprintf() function.
   * CRTDebug.c: changed the vsnprintf() usage to use vasprintf() instead.
     This removes the 1024 character limit for outputting a string.

2007-03-19  Jens Langner <Jens.Langner@light-speed.de>

   * CRTDebug.c: added DBC_TIMEVAL to the default debugging classes in case
     no debugging variable is found during init.

2007-03-07  Jens Langner <Jens.Langner@light-speed.de>

   * CRTDebug.c: exchanged the colors for DEBUG and REPORT output. DEBUG
     output should now be green and reports be outputted in purple instead.

2006-11-30  Jens Langner <Jens.Langner@light-speed.de>

	* CRTDebug.c: Now the time calculations should finally be correct. I
     really can't believe how easy it is/was to mix up milli with micro :)
	* CRTDebug.c: fixed the STARTCLOCK()/STOPCLOCK() functionality as it was
	  outputting incorrect time values which where quiet confusing.

2006-09-08  Jens Langner <Jens.Langner@light-speed.de>

	* examples: added new examples directory for preparing some small, minor
     examples on how to use rtdebug within own projects.

#### 2.0 RELEASE ##########################################################

2006-08-28  Jens Langner <Jens.Langner@light-speed.de>

	* CRTDebug.cpp,h: moved all private data into an internal-only private
		inline class "CRTDebugPrivate" which we might enhance in future
		without having to expose it to the public.

2006-08-25  Jens Langner <Jens.Langner@light-speed.de>

	* misc: version bump due to changed library interface
	* CRTDebug.cpp: we use GetTickCount() on systems not supporting
		gettimeofday().

2006-08-18  Jens Langner <Jens.Langner@light-speed.de>

	* CRTDebug.cpp: in case no pthread library is found librtdebug will
		now be compiled in "single-threaded" mode instead of refusing to
		compile at all.
	* misc: reworked the whole debugging interface to load its debugging
	  classes/flags via a user-defineable ENV variable. In addition, a
		filename matching facility has been added to make it also possible
		to specify a string pattern which, in case it matches, will
		suppress/show the debugging output in question.
		Now all debug flags/classes are fully configurable/supressable. In
		addition source files can be combined into so-called "debug modules"
		which can also be suppressed or explicitly showed.

#### 1.0 RELEASE ##########################################################

2005-01-27  Jens Langner <Jens.Langner@light-speed.de>

  * first internal 1.0 release


src/include/rtdebug.h
/* vim:set ts=2 nowrap: ****************************************************

 librtdebug - A C++ based thread-safe Runtime Debugging Library
 Copyright (C) 2003-2006 by Jens Langner <Jens.Langner@light-speed.de>

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 $Id$

***************************************************************************/

#ifndef RTDEBUG_H
#define RTDEBUG_H

//! This header file is part of the CDebug class mechanism to allow
//! developers to output debugging relevant information to a terminal
//! with automatically taking respect of different a multithreaded
//! environment.
//!
//! Please take a look at the CRTDebug.h header file for a more
//! detailed information about the ideas behind that debugging mechanism

// redefine the Qt's own debug system
#ifdef qDebug
#undef qDebug
#define qDebug	D
#endif
#ifdef qWarning
#undef qWarning
#define qWarning W
#endif
#ifdef qFatal
#undef qFatal
#define qFatal E
#endif

#include <CRTDebug.h>

// first we make sure all previously defined symbols are undefined now so
// that no other debug system interferes with ours.
#if defined(ENTER)
#undef ENTER
#endif
#if defined(LEAVE)
#undef LEAVE
#endif
#if defined(RETURN)
#undef RETURN
#endif
#if defined(SHOWVALUE)
#undef SHOWVALUE
#endif
#if defined(SHOWPOINTER)
#undef SHOWPOINTER
#endif
#if defined(SHOWSTRING)
#undef SHOWSTRING
#endif
#if defined(SHOWMSG)
#undef SHOWMSG
#endif
#if defined(STARTCLOCK)
#undef STARTCLOCK
#endif
#if defined(STOPCLOCK)
#undef STOPCLOCK
#endif
#if defined(D)
#undef D
#endif
#if defined(E)
#undef E
#endif
#if defined(W)
#undef W
#endif
#if defined(ASSERT)
#undef ASSERT
#endif
#if defined(Debug)
#undef Debug
#endif
#if defined(Verbose)
#undef Verbose
#endif
#if defined(Warning)
#undef Warning
#endif
#if defined(Error)
#undef Error
#endif
#if defined(Info)
#undef Info
#endif

#if defined(DEBUG)

#include <stdlib.h>

#if !defined(DEBUG_MODULE)
#define DEBUG_MODULE DBM_NONE
#endif

// Core class information class messages
#define ENTER()					CRTDebug::instance()->Enter(DBC_CTRACE, DEBUG_MODULE, __FILE__, __LINE__, __FUNCTION__)
#define LEAVE()					CRTDebug::instance()->Leave(DBC_CTRACE, DEBUG_MODULE, __FILE__, __LINE__, __FUNCTION__)
#define RETURN(r)				CRTDebug::instance()->Return(DBC_CTRACE, DEBUG_MODULE, __FILE__, __LINE__, __FUNCTION__, (long)r)
#define SHOWVALUE(v)		CRTDebug::instance()->ShowValue(DBC_REPORT, DEBUG_MODULE, (long long)v, sizeof(v), #v, __FILE__, __LINE__)
#define SHOWPOINTER(p)	CRTDebug::instance()->ShowPointer(DBC_REPORT, DEBUG_MODULE, p, #p, __FILE__, __LINE__)
#define SHOWSTRING(s)		CRTDebug::instance()->ShowString(DBC_REPORT, DEBUG_MODULE, s, #s, __FILE__, __LINE__)
#define SHOWMSG(m)			CRTDebug::instance()->ShowMessage(DBC_REPORT, DEBUG_MODULE, m, __FILE__, __LINE__)
#define STARTCLOCK(s)		CRTDebug::instance()->StartClock(DBC_TIMEVAL, DEBUG_MODULE,  s, __FILE__, __LINE__)
#define STOPCLOCK(s)		CRTDebug::instance()->StopClock(DBC_TIMEVAL, DEBUG_MODULE, s, __FILE__, __LINE__)
#define D(s, vargs...)	CRTDebug::instance()->dprintf_header(DBC_DEBUG, DEBUG_MODULE, __FILE__, __LINE__, s, ## vargs)
#define E(s, vargs...)	CRTDebug::instance()->dprintf_header(DBC_ERROR, DEBUG_MODULE, __FILE__, __LINE__, s, ## vargs)
#define W(s, vargs...)	CRTDebug::instance()->dprintf_header(DBC_WARNING, DEBUG_MODULE, __FILE__, __LINE__, s, ## vargs)
#define ASSERT(expression)      \
	((void)                       \
	 ((expression) ? 0 :          \
	  (														\
	   CRTDebug::instance()->dprintf_header(DBC_ASSERT,		\
																					DEBUG_MODULE, \
																					__FILE__,			\
																					__LINE__,			\
																					"failed assertion '%s'", #expression), \
	   abort(),                   \
	   0                          \
	  )                           \
	 )                            \
	)

#else // DEBUG

#define ENTER()							(void(0))
#define LEAVE()							(void(0))
#define RETURN(r)						(void(0))
#define SHOWVALUE(v)				(void(0))
#define SHOWPOINTER(p)			(void(0))
#define SHOWSTRING(s)				(void(0))
#define SHOWMSG(m)					(void(0))
#define STARTCLOCK(s)				(void(0))
#define STOPCLOCK(s)				(void(0))
#define D(s, vargs...)			(void(0))
#define E(s, vargs...)			(void(0))
#define W(s, vargs...)			(void(0))
#define ASSERT(expression)	(void(0))

#endif // DEBUG
#endif // RTDEBUG_H


examples/hello_world.cpp
/* vim:set ts=2 nowrap: ****************************************************

 librtdebug - A C++ based thread-safe Runtime Debugging Library
 Copyright (C) 2003-2006 by Jens Langner <Jens.Langner@light-speed.de>

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 $Id$

***************************************************************************/

/*
 * This is a small hello world example on how to use librtdebug which
 * should demonstrate how easy it is to integrate rtdebug into everyone's
 * projects. By using just a small set of preprocessor macros (defined in
 * rtdebug.h) a user can easily setup his program for outputting very
 * detailed and sophisticated debugging output.
 *
 * And by using an environment variable, this output is then even dynamically 
 * selectable upon startup of the application.
 *
 * So all I can say it, use "rtdebug" and have a happy debugging session time!
 *
 * cheers,
 * jens
 *
 */

#include <iostream>

// rtdebug main include which will define all our
// preprocessor macros
#include <rtdebug.h>

using namespace std;

// prototypes
void output(const char* text);

// main function
int main(int argc, char* argv[])
{
	int returnCode = EXIT_SUCCESS; // return no error on default

	// for being able to filter the debugging output via
	// an environment variable, a user can use ::init() to
	// define that ENV-variable.
	#if defined(DEBUG)
	CRTDebug::init("hello_world");
	#endif

	// we delay the first function entry marker because of ::init()
	ENTER();

	// we measure the time of the execution from here on
	STARTCLOCK("output() measurement");

	// now we branch into the output() function
	output("Hello to rtDebug!");

	// signal that we are finished
	STOPCLOCK("output() measurement");

	// before we really exit the main function we have to
	// use RETURN() to report the return value to the debug framework
	RETURN(returnCode);

	// for properly cleaning up the debug environment we have to
	// call ::destroy() as the very last method of CRTDebug
	#if defined(DEBUG)
	CRTDebug::destroy();
	#endif

	return returnCode;
}

void output(const char* text)
{
	ENTER();
	bool result = true;

	// for debug purposes we can output the content of a c string
	// via SHOWSTRING()
	SHOWSTRING(text);

	cout << text << endl;

	LEAVE();
}


src/CRTDebug.h
/* vim:set ts=2 nowrap: ****************************************************

 librtdebug - A C++ based thread-safe Runtime Debugging Library
 Copyright (C) 2003-2006 by Jens Langner <Jens.Langner@light-speed.de>

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.

 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.

 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 $Id$

***************************************************************************/

#ifndef CRTDEBUG_H
#define CRTDEBUG_H

// debug classes
#define DBC_CTRACE		(1<<0) // call tracing (ENTER/LEAVE etc.)
#define DBC_REPORT		(1<<1) // reports (SHOWVALUE/SHOWSTRING etc.)
#define DBC_ASSERT		(1<<2) // asserts (ASSERT)
#define DBC_TIMEVAL		(1<<3) // time evaluations (STARTCLOCK/STOPCLOCK)
#define DBC_DEBUG			(1<<4) // debugging output D()
#define DBC_ERROR			(1<<5) // error output     E()
#define DBC_WARNING		(1<<6) // warning output   W()
#define DBC_ALL				(0xffffffff)

// debug flags
#define DBF_ALWAYS		(1<<0)
#define DBF_STARTUP		(1<<1)
#define DBF_ALL				(0xffffffff)

// debug modules
#define DBM_NONE			NULL
#define DBM_ALL				"all"

// forward declarations
class CRTDebugPrivate;

//  Classname:   CRTDebug
//! @brief debugging purpose class
//! @ingroup debug
//!
//! This class manages the whole debugging system of an application. It will
//! print out specified debugging information depending on the currently set
//! debugging level.
//! To use this class properly it is necessary that you specify your own
//! debugging levels first (specify the above class defines and the blow enum Levels)
//! so that you can specify which debug message belongs to which class and the
//! below "enum Level" should contain the different bit combinations of those
//! debug "classes", so that you can set a different debug level on runtime.
//!
//! Last, but not least you need to include the file "debug.h" everywhere you
//! want to put debugmessages which directly uses this debug class. Always use
//! the predefined macros in "debug.h" for specifying your debug messages.
//! There you will fine uppercase-only macros which will automatically removed
//! when you compile your program without the global "DEBUG" define, and you got
//! mixed-case macros which will stay in your code, but behave differently if
//! compiled without the "DEBUG" define.
//! Normally the mixed-case macros are used for terminal output which should
//! stay and be user configurable and should not be placed in time consuming
//! situations. The uppercase macros are really for debugging purposes only and
//! will completely vanish if you build a release version.
//! The rule is: Use the uppercase macros to print out your geek debug
//!              information so that if you compile a release version a normal
//!              user wouldn't be bothered with. The mixed-case macros will
//!              output user-dependant debug information.
//!
//! Please note that this debug class is fully threadsafe - which means that
//! it will take care of overlapping output from different threads. It will
//! sync it with a mutal exclusion and as an nice addition also output the
//! threadnumber in front of every debug output to easily identify to which
//! thread a output belongs to.
////////////////////////////////////////////////////////////////////////////////
class CRTDebug
{
	public:
		// the static singleton instance method
		static CRTDebug* instance();
		static void destroy();

		// for initialization via ENV variables
		static void init(const char* variable=0);

		// our main debug output methods
		void Enter(const int c, const char* m, const char* file, long line, const char* function);
		void Leave(const int c, const char* m, const char* file, int line, const char* function);
		void Return(const int c, const char* m, const char* file, int line, const char* function, long result);
		void ShowValue(const int c, const char* m, long long value, int size, const char* name, const char* file, long line);
		void ShowPointer(const int c, const char* m, void* pointer, const char* name, const char* file, long line);
		void ShowString(const int c, const char* m, const char* string, const char* name, const char* file, long line);
		void ShowMessage(const int c, const char* m, const char* string, const char* file, long line);
		void StartClock(const int c, const char* m, const char* string, const char* file, long line);
		void StopClock(const int c, const char* m, const char* string, const char* file, long line);

		// some raw methods to format text like printf() does
		void dprintf_header(const int c, const char* m, const char* file, long line, const char* fmt, ...);
		void dprintf(const int c, const char* m, const char* fmt, ...);

		// general public methods to control debug class
		unsigned int debugClasses() const;
		unsigned int debugFlags() const;
		const char* debugFiles() const;
		const char* debugModules() const;
		void setDebugClass(unsigned int cl);
		void setDebugFlag(unsigned int fl);
		void setDebugFile(const char* filename, bool show);
		void setDebugModule(const char* module, bool show);
		void clearDebugClass(unsigned int cl);
		void clearDebugFlag(unsigned int fl);
		void clearDebugFile(const char* filename);
		void clearDebugModule(const char* module);

		// methods to control additional options
		bool highlighting() const;
		void setHighlighting(bool on);

	protected:
		CRTDebug(const int dbclasses=0, const int dbflags=0);
		~CRTDebug();

	private:
		static CRTDebug*	m_pSingletonInstance;	//!< the singleton instance
		CRTDebugPrivate*	m_pData;							//!< the private, internal rtdebug data
};

#endif // CRTDEBUG_H
